include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl_utils.h

```diff
@@ -412,68 +412,71 @@ using __repacked_tuple_t = typename __repacked_tuple<T>::type;
 template <typename _ContainerOrIterable>
 using __value_t = typename __internal::__memobj_traits<_ContainerOrIterable>::value_type;
 
-//-----------------------------------------------------------------------
-// future and helper classes for async pattern/algorithm
-//-----------------------------------------------------------------------
+template <typename _Event>
+void __wait_event(_Event);
 
-// TODO: towards higher abstraction and generic future. implementation specific sycl::event should be hidden
-struct __future_base
+void
+__wait_event(sycl::event __e)
 {
-    sycl::event __my_event;
-
-    __future_base() = default;
-    __future_base(sycl::event __e) : __my_event(__e) {}
-    void
-    wait()
-    {
 #if !ONEDPL_ALLOW_DEFERRED_WAITING
-        __my_event.wait_and_throw();
+    __e.wait_and_throw();
 #endif
-    }
-    operator sycl::event() const { return __my_event; }
-};
+}
 
 template <typename _T>
-class __future : public __future_base
+constexpr auto
+__get_value(sycl::buffer<_T>& __buf)
 {
-    ::std::size_t __result_idx;
-    sycl::buffer<_T> __data;
-
-  public:
-    __future(sycl::event __e, size_t __o, sycl::buffer<_T> __b)
-        : __par_backend_hetero::__future_base(__e), __result_idx(__o), __data(__b)
-    {
-    }
+    //according to a contract, returned value is one-element sycl::buffer
+    return __buf.template get_access<access_mode::read>()[0];
+}
 
-    _T
-    get()
-    {
-        return __data.template get_access<access_mode::read>()[__result_idx];
-    }
-    template <class _Tp, class _Enable>
-    friend class oneapi::dpl::__internal::__future;
-};
+template <typename _T>
+constexpr auto
+__get_value(_T& __val)
+{
+    return __val;
+}
 
-template <>
-class __future<void> : public __future_base
+//A contract for future class: <sycl::event or other event, a value or sycl::buffers...>
+//Impl details: inheretance (private) instead of aggregation for enabling the empty base optimization.
+template <typename _Event, typename... Args>
+class __future : private std::tuple<Args...>
 {
-    ::std::unique_ptr<oneapi::dpl::__internal::__lifetime_keeper_base> __tmps;
+    _Event __my_event;
+    static constexpr bool __is_value = sizeof...(Args) > 0;
 
   public:
-    template <typename... _Ts>
-    __future(sycl::event __e, _Ts... __t) : __future_base(__e)
+    __future(_Event e, Args... args) : __my_event(e), std::tuple<Args...>(args...) {}
+    __future(_Event e, std::tuple<Args...> _t) : __my_event(e), std::tuple<Args...>(_t) {}
+
+    operator _Event() const { return __my_event; }
+    auto
+    event() const
     {
-        if (sizeof...(__t) != 0)
-            __tmps = ::std::unique_ptr<oneapi::dpl::__internal::__lifetime_keeper<_Ts...>>(
-                new oneapi::dpl::__internal::__lifetime_keeper<_Ts...>(__t...));
+        return operator _Event();
     }
     void
+    wait()
+    {
+        __wait_event(operator _Event());
+    }
+
+    template <typename = std::enable_if<__is_value, void>>
```

Comment:
Cool. We have the expression that is always well-formed (it's not `enable_if_t`). Thus, it forces me to ask again do we really think that kind of "constaint" really helpful? Seems like you don't have enough of testing but on the other hand all the tests are passed (I hope). This is one of the first candidates to reconsider (to me).

Reply:
I do think that constraining `get()` to be only available for non-void futures is appropriate, if done correctly of course - because I think calling `get()` for an event-only future should be considered an error. 

Do you think this capability is somehow useful (to keep some code generic, maybe?), and if yes - can you suggest a practical use case?

Reply:
Essentially, if we have just one "submit" and got a write access for sycl::buffer where a result to be written, a dependency graph built by SYCL Runtime provides blocking on getting access on the host for the mentioned buffer.
But, if we have several "submits" and make a synchronization  based on sycl::event dependency, yes, we have to explicit call wait on the event of future object.

Reply:
> (it's not enable_if_t)

But, `std::enable_if` is a C++ type as well. And such approach works, have a look: https://godbolt.org/z/dc5cd7nsd

Reply:
While added `//TODO: to add SFINAE conditional for a case when sizeof...(_Args) = 0`...
In general, I agree with Alexey, "that constraining get() to be only available for non-void futures is appropriate".